futures-sink/src/lib.rs

```diff
@@ -172,21 +62,21 @@ pub trait Sink {
 
     /// Attempts to prepare the `Sink` to receive a value.
     ///
-    /// This method must be called and return `Ok(Async::Ready(()))` prior to
+    /// This method must be called and return `Ok(Poll::Ready(()))` prior to
     /// each call to `start_send`.
     ///
-    /// This method returns `Async::Ready` once the underlying sink is ready to
+    /// This method returns `Poll::Ready` once the underlying sink is ready to
     /// receive data. If this method returns `Async::Pending`, the current task
     /// is registered to be notified (via `cx.waker()`) when `poll_ready`
     /// should be called again.
     ///
     /// In most cases, if the sink encounters an error, the sink will
     /// permanently be unable to receive items.
-    fn poll_ready(&mut self, cx: &mut task::Context) -> Poll<(), Self::SinkError>;
+    fn poll_ready(self: PinMut<Self>, cx: &mut task::Context) -> Poll<Result<(), Self::SinkError>>;
```

Comment:
Rationale on keeping the error baked in here?

Reply:
It doesn't necessarily need to be an error, I suppose-- we could instead return an associated `type Result`. I think that'd always be either `()` or `Result`, in most cases it would be `Result`, and keeping it as `Result` allows for having consistently short-circuiting operations, and avoids questions about what to do with the value when making a future that calls multiple poll functions and receives multiple `Sink::Result`s, like the `Send` combinator. I think I could definitely be convinced that it should change to an associated type, but I don't see too strong of a motivation, and the short-circuiting behavior of several combinators is useful.